File: domain/trackers.py
"""domain/trackers.py -- watch a number on a listing and say when it moves.

    Orbit has four of these in its menu: BSR Tracker, BuyBox Tracker, Price
    Tracker, Fee Tracker, plus "All Trackers" over the top and one Alerts count
    fed by all of them.

FOUR TRACKERS, ONE ENGINE.

Looked at closely, the four are the same thing pointed at different numbers.
Each is a per-ASIN watch list holding: the value now, the value you want, how
far it has drifted, when it was last read, and whether you are watching it at
all. Four modules would be four copies of that logic drifting apart, and the
first thing to drift would be the alert rule -- so there is one engine here and
the metrics are DATA (see METRICS below), not code. CLAUDE.md Rule 12.

Adding a fifth tracker means adding a row to METRICS and a fetcher. It does not
mean touching the store, the drift maths, the alert rule, the route or the
screen.

WHAT IS STORED, AND WHAT IS NOT.

Deliberately NOT stored: the list of products. That already exists in the
account's catalogue, and a second copy would go stale the moment a listing was
added -- the classic symptom being a new product that can never be tracked
because it is not in the tracker's own list. This module stores only what the
catalogue cannot know:

    watch[asin][metric] = {"on": bool, "target": float|None}
    history[asin][metric] = [{"at": iso, "v": float}, ...]   newest last

DRIFT IS AGAINST WHAT YOU ASKED FOR, NOT AGAINST YESTERDAY.

The number that matters is distance from target, because that is the thing you
can act on. Movement since the last reading is also reported (`change`) but it
is not what raises an alert: a rank that has been 40% worse than target for a
month is a problem every day of that month, and an alert that fires only on the
day it moves would have gone quiet after the first one.

A MISSING READING IS NOT A ZERO.

If Amazon does not answer for an ASIN, its value is None and its status is
UNKNOWN. It is never 0, never "improved to nothing", and it never counts towards
the alert total. Reporting a failed fetch as a perfect score is the single
easiest way to make a monitoring screen lie, and this app has already been bitten
by exactly that shape of bug in the stock and buy-box readers.
"""
import datetime
import threading

from domain import jsonstore

_LOCK = threading.Lock()
_FILE = "trackers.json"

# How many readings to keep per ASIN per metric. Enough to draw a trend and to
# survive a week of hourly checks; small enough that the file stays a file.
MAX_HISTORY = 400

# WHICH WAY IS GOOD.
#
# The single most important field here. A sales rank of 900 is BETTER than
# 4,000; a price of 9.99 is worse than 12.99 if you are trying to hold a price.
# Getting this backwards makes every alert exactly wrong, and it is invisible
# from the numbers alone -- which is why it is written down per metric rather
# than inferred anywhere.
LOWER_IS_BETTER = "lower"
HIGHER_IS_BETTER = "higher"

# `kind` drives formatting only: "money" gets the marketplace's currency symbol,
# "rank" gets thousands separators and a #, "percent" gets a %.
METRICS = {
    "bsr": {
        "label": "Sales rank",
        "tracker": "BSR Tracker",
        "kind": "rank",
        "good": LOWER_IS_BETTER,
        # 20% off target is a real move in rank terms; rank is noisy day to day.
        "tolerance": 0.20,
        "blurb": "Where this product sits in its category. Lower is better.",
        "source": "Amazon's catalogue (sales ranks)",
    },
    "buybox": {
        "label": "Buy Box price",
        "tracker": "BuyBox Tracker",
        "kind": "money",
        "good": LOWER_IS_BETTER,
        "tolerance": 0.05,
        "blurb": "What the Buy Box is currently priced at, whoever holds it.",
        "source": "Amazon's live offers",
    },
    "price": {
        "label": "Your price",
        "tracker": "Price Tracker",
        "kind": "money",
        "good": HIGHER_IS_BETTER,
        "tolerance": 0.05,
        "blurb": "What you are charging. Drifts when a repricer moves it.",
        "source": "Amazon's live offers",
    },
    "fee": {
        "label": "Amazon's cut",
        "tracker": "Fee Tracker",
        "kind": "money",
        "good": LOWER_IS_BETTER,
        "tolerance": 0.10,
        "blurb": "The referral fee Amazon quotes on this ASIN at your price.",
        "source": "Amazon's fee estimate",
    },
}

OK = "ok"
OFF = "off"
UNKNOWN = "unknown"


def _now():
    return datetime.datetime.now().strftime("%Y-%m-%d %H:%M")


def _num(v):
    """A float, or None. None and 'not a number' are the same answer: unknown."""
    if v is None or v is True or v is False:
        return None
    try:
        f = float(v)
    except (TypeError, ValueError):
        return None
    # NaN fails this comparison with itself, which is the cheapest way to catch it.
    return f if f == f else None


def _path(config_path):
    return jsonstore.path_beside_config(config_path, _FILE)


def _blank():
    return {"watch": {}, "history": {}}


def load(config_path):
    d = jsonstore.read_json(_path(config_path), None)
    if not isinstance(d, dict):
        return _blank()
    if not isinstance(d.get("watch"), dict):
        d["watch"] = {}
    if not isinstance(d.get("history"), dict):
        d["history"] = {}
    return d


def _save(config_path, data):
    return jsonstore.write_json_atomic(_path(config_path), data, indent=2)


def _wskey(workspace_id, asin):
    """One key per account per ASIN.

    The account is part of the key because the same ASIN can be listed by two of
    these accounts at different prices, and one shared row would have them
    overwriting each other's readings -- which reads as a price that flaps
    between two values for no reason.
    """
    return "%s::%s" % (str(workspace_id or "").strip(), str(asin or "").strip().upper())


# --------------------------------------------------------------------------
# the watch list
# --------------------------------------------------------------------------

def watch_get(config_path, workspace_id, asin, metric):
    """{"on":bool, "target":float|None} for one ASIN and one metric."""
    row = (load(config_path).get("watch") or {}).get(_wskey(workspace_id, asin)) or {}
    e = row.get(metric) or {}
    return {"on": bool(e.get("on")), "target": _num(e.get("target"))}


def watch_set(config_path, workspace_id, asin, metric, on=None, target=None):
    """Turn a tracker on or off for one ASIN, and/or set its target.

    `on` and `target` are both optional so a screen can change one without
    having to know the other -- passing None for `target` would otherwise be
    indistinguishable from asking to clear it. To clear a target, pass the
    string "" (which _num turns into None) rather than omitting it.
    """
    if metric not in METRICS:
        return {"ok": False, "error": "Unknown tracker: %s" % metric}
    k = _wskey(workspace_id, asin)
    with _LOCK:
        data = load(config_path)
        row = data.setdefault("watch", {}).setdefault(k, {})
        e = row.setdefault(metric, {"on": False, "target": None})
        if on is not None:
            e["on"] = bool(on)
        if target is not None:
            e["target"] = _num(target)
        _save(config_path, data)
        return {"ok": True, "watch": {"on": bool(e.get("on")),
                                      "target": _num(e.get("target"))}}


def tracked(config_path, workspace_id, metric=None):
    """The ASINs being watched, as {asin: {metric: {...}}}.

    Only rows with at least one metric switched ON, because an ASIN whose target
    was set and then untracked should not be fetched. Fetching costs an API call
    per ASIN and this is the list that decides how many.
    """
    out = {}
    prefix = "%s::" % str(workspace_id or "").strip()
    for k, row in (load(config_path).get("watch") or {}).items():
        if not k.startswith(prefix):
            continue
        asin = k[len(prefix):]
        keep = {}
        for m, e in (row or {}).items():
            if m in METRICS and (e or {}).get("on"):
                if metric and m != metric:
                    continue
                keep[m] = {"on": True, "target": _num(e.get("target"))}
        if keep:
            out[asin] = keep
    return out


# --------------------------------------------------------------------------
# readings
# --------------------------------------------------------------------------

def record(config_path, workspace_id, asin, metric, value, at=None):
    """Append one reading. A value that is not a number is NOT stored.

    Refusing to store the unknown is what keeps a failed fetch from becoming a
    data point. If Amazon does not answer, the last real reading stays the last
    real reading and its age is visible, which is the honest picture; writing a
    None into the history would instead show a line dropping to the floor.
    """
    v = _num(value)
    if metric not in METRICS or v is None:
        return False
    k = _wskey(workspace_id, asin)
    with _LOCK:
        data = load(config_path)
        hist = data.setdefault("history", {}).setdefault(k, {}).setdefault(metric, [])
        hist.append({"at": at or _now(), "v": v})
        if len(hist) > MAX_HISTORY:
            del hist[:len(hist) - MAX_HISTORY]
        _save(config_path, data)
    return True


def history(config_path, workspace_id, asin, metric, limit=0):
    h = ((load(config_path).get("history") or {})
         .get(_wskey(workspace_id, asin)) or {}).get(metric) or []
    return h[-limit:] if limit and limit > 0 else list(h)


def latest(config_path, workspace_id, asin, metric):
    h = history(config_path, workspace_id, asin, metric)
    return h[-1] if h else None


# --------------------------------------------------------------------------
# the judgement
# --------------------------------------------------------------------------

def drift(value, target, good):
    """How far off target, as a signed fraction. None when it cannot be known.

    Positive means WORSE than target, negative means better, whichever direction
    "better" happens to be for this metric. Every caller can then use one rule --
    positive is bad -- instead of each remembering which way round its metric is,
    which is precisely the thing that would get it backwards somewhere.
    """
    v, t = _num(value), _num(target)
    if v is None or t is None or t == 0:
        return None
    raw = (v - t) / abs(t)
    return raw if good == LOWER_IS_BETTER else -raw


def status_for(value, target, good, tolerance):
    """OK / OFF / UNKNOWN for one reading.

    UNKNOWN covers two different situations on purpose: no reading, and no
    target. Neither is a pass and neither is a failure -- an ASIN you have not
    given a target cannot be off track, and saying OK would claim a judgement
    that was never made.
    """
    d = drift(value, target, good)
    if d is None:
        return UNKNOWN
    return OFF if d > (tolerance if tolerance is not None else 0) else OK


def rows(config_path, workspace_id, metric=None, names=None):
    """One row per tracked ASIN per metric -- what the screen draws.

    `names` is an optional {asin: title} so the screen can show a product name
    without this module having to know how to look one up. Kept as a parameter
    rather than a lookup because domain/catalogue.py already owns that job.
    """
    out = []
    watch = tracked(config_path, workspace_id, metric)
    for asin, metrics in sorted(watch.items()):
        for m, e in sorted(metrics.items()):
            spec = METRICS[m]
            last = latest(config_path, workspace_id, asin, m)
            h = history(config_path, workspace_id, asin, m, limit=2)
            prev = h[0]["v"] if len(h) > 1 else None
            v = last["v"] if last else None
            tgt = e.get("target")
            d = drift(v, tgt, spec["good"])
            out.append({
                "asin": asin,
                "name": (names or {}).get(asin, ""),
                "metric": m,
                "label": spec["label"],
                "tracker": spec["tracker"],
                "kind": spec["kind"],
                "good": spec["good"],
                "value": v,
                "target": tgt,
                "drift": d,
                # Movement since the previous reading. Reported, but never the
                # thing that raises an alert -- see the module docstring.
                "change": (None if (v is None or prev is None) else round(v - prev, 4)),
                "last_at": last["at"] if last else "",
                "points": len(history(config_path, workspace_id, asin, m)),
                "status": status_for(v, tgt, spec["good"], spec["tolerance"]),
            })
    return out


def alerts(config_path, workspace_id, names=None):
    """Every row that is OFF target, worst first, plus the count.

    ONE count across all four trackers, as Orbit has it. A per-tracker count
    would mean four badges to check, which is four chances to miss the one that
    mattered.
    """
    bad = [r for r in rows(config_path, workspace_id, None, names)
           if r["status"] == OFF]
    bad.sort(key=lambda r: (-(r["drift"] or 0), r["asin"]))
    return {"count": len(bad), "rows": bad}


def summary(config_path, workspace_id, names=None):
    """Counts per tracker for the All Trackers screen."""
    all_rows = rows(config_path, workspace_id, None, names)
    out = {}
    for m, spec in METRICS.items():
        mine = [r for r in all_rows if r["metric"] == m]
        out[m] = {
            "metric": m,
            "tracker": spec["tracker"],
            "label": spec["label"],
            "blurb": spec["blurb"],
            "source": spec["source"],
            "tracked": len(mine),
            "off": len([r for r in mine if r["status"] == OFF]),
            "ok": len([r for r in mine if r["status"] == OK]),
            "unknown": len([r for r in mine if r["status"] == UNKNOWN]),
        }
    return out

File: static/js/trackers.js
// static/js/trackers.js — the four trackers and the alert list.
//
//     Orbit's menu has All Trackers, BSR Tracker, BuyBox Tracker, Price Tracker,
//     Fee Tracker and Alerts.
//
// One screen with a tab each. The four are the same table pointed at a different
// number, so five separate screens would be four copies of one table drifting
// apart — and the first thing to drift would be how each one decides that a row
// is off target, which is the only judgement the screen makes.
//
// The engine (domain/trackers.py) does all of that judging. This file draws what
// it is told and never re-derives it: no thresholds here, no "which way is good"
// here. That matters because a sales rank of 900 beats 4,000 while a price of
// 9.99 does not beat 12.99, and a second opinion about which is which — held in
// JavaScript, invisible from the Python — is exactly the bug nobody would find.

let TRK = { metric: "", metrics: {}, rows: [], loading: false };

// Which account and marketplace this screen is about. Through the one shared
// builder rather than a fifth hand-rolled copy -- see static/js/scopeq.js for
// the two real faults that comparing the previous four turned up.
function _trkQs() { return (typeof scopeQs === "function") ? scopeQs() : ""; }

function trkFmt(v, kind, cur) {
  if (v === null || v === undefined || v === "") return "—";
  const n = Number(v);
  if (isNaN(n)) return "—";
  if (kind === "money") return (cur || "") + n.toFixed(2);
  if (kind === "rank") return "#" + Math.round(n).toLocaleString();
  if (kind === "percent") return n.toFixed(1) + "%";
  return String(n);
}

// Drift is already signed so that POSITIVE MEANS WORSE for every metric,
// whichever direction better happens to be. Reading it any other way here would
// re-introduce the per-metric question the engine exists to have answered once.
function trkDrift(d) {
  if (d === null || d === undefined) return '<span class="cc">—</span>';
  const pct = (d * 100);
  const cls = d > 0 ? "trkbad" : "trkgood";
  const sign = d > 0 ? "+" : "";
  return '<span class="' + cls + '">' + sign + pct.toFixed(1) + "%</span>";
}

function trkStatus(s) {
  if (s === "off") return '<span class="trkpill off">Off target</span>';
  if (s === "ok") return '<span class="trkpill ok">On target</span>';
  // NOT a tick and NOT a cross. "Could not be read" and "no target set" are
  // both genuinely unknown, and showing either as a pass is how a monitoring
  // screen starts lying.
  return '<span class="trkpill unk">Not known</span>';
}

function trkCur() {
  return (typeof curSymbol === "function") ? curSymbol("") : "";
}

function trkTabs() {
  const box = document.getElementById("trk_tabs");
  if (!box) return;
  const ms = TRK.metrics || {};
  let html = '<div class="stk-tab' + (TRK.metric ? "" : " on") +
             '" onclick="trkTab(\'\')">All trackers</div>';
  Object.keys(ms).forEach(function (k) {
    const m = ms[k];
    const on = (TRK.metric === k) ? " on" : "";
    html += '<div class="stk-tab' + on + '" onclick="trkTab(\'' + k + '\')">' +
            esc(m.tracker || k) + "</div>";
  });
  box.innerHTML = html;
}

function trkTab(m) {
  TRK.metric = m || "";
  trkTabs();
  trkRender();
}

function trkRender() {
  const box = document.getElementById("trk_body");
  if (!box) return;
  const cur = trkCur();
  const rows = TRK.metric ? TRK.rows.filter(function (r) { return r.metric === TRK.metric; })
                          : TRK.rows;
  if (!rows.length) {
    // An empty screen has to say what to DO, not just that it is empty. The
    // trackers are opt-in per ASIN, so "nothing here" is the normal first state
    // and needs to read as a starting point rather than a failure.
    box.innerHTML =
      '<div class="card" style="padding:18px">' +
      '<div style="font-weight:600;margin-bottom:6px">Nothing is being tracked yet</div>' +
      '<div class="cc" style="font-size:12.5px;line-height:1.55;max-width:640px">' +
      'A tracker watches one number on one listing and tells you when it moves away ' +
      'from what you wanted. Add an ASIN below, choose which number to watch and what ' +
      'you are aiming for, then press <b>Check now</b>. Nothing is read from Amazon ' +
      'until you do — each check costs an API call, so it happens when you ask.' +
      "</div>" + trkAddForm() + "</div>";
    return;
  }
  let html = trkAddForm() +
    '<div class="card" style="overflow-x:auto"><table class="stk-table"><thead><tr>' +
    "<th>Product</th><th>Tracker</th><th>Now</th><th>Target</th><th>Drift</th>" +
    "<th>Change</th><th>Status</th><th>Last read</th><th></th>" +
    "</tr></thead><tbody>";
  rows.forEach(function (r) {
    html += "<tr>" +
      "<td><div style=\"font-weight:600\">" + esc(r.asin) + "</div>" +
      '<div class="cc" style="font-size:11px">' + esc((r.name || "").slice(0, 60)) + "</div></td>" +
      "<td>" + esc(r.tracker) + "</td>" +
      "<td>" + trkFmt(r.value, r.kind, cur) + "</td>" +
      '<td><input class="ed trktgt" style="width:88px;padding:3px 6px;font-size:12px" ' +
      'value="' + (r.target === null || r.target === undefined ? "" : r.target) + '" ' +
      "onchange=\"trkSetTarget('" + r.asin + "','" + r.metric + "',this.value)\"></td>" +
      "<td>" + trkDrift(r.drift) + "</td>" +
      "<td>" + (r.change === null || r.change === undefined ? '<span class="cc">—</span>'
                                                           : trkFmt(r.change, r.kind, cur)) + "</td>" +
      "<td>" + trkStatus(r.status) + "</td>" +
      '<td class="cc" style="font-size:11px">' + esc(r.last_at || "never") +
      (r.points ? ' <span class="cc">(' + r.points + ")</span>" : "") + "</td>" +
      '<td><button class="ib" title="Stop tracking this" onclick="trkStop(\'' +
      r.asin + "','" + r.metric + "')\"><i class=\"ti ti-x\"></i></button></td>" +
      "</tr>";
  });
  html += "</tbody></table></div>";
  box.innerHTML = html;
}

function trkAddForm() {
  const ms = TRK.metrics || {};
  let opts = "";
  Object.keys(ms).forEach(function (k) {
    opts += '<option value="' + k + '">' + esc(ms[k].label || k) + "</option>";
  });
  return '<div class="card" style="padding:12px;margin-bottom:10px;display:flex;' +
    'gap:8px;align-items:flex-end;flex-wrap:wrap">' +
    '<div><label class="cc" style="display:block;font-size:11px">ASIN</label>' +
    '<input id="trk_asin" class="ed" style="width:130px" placeholder="B0XXXXXXXX"></div>' +
    '<div><label class="cc" style="display:block;font-size:11px">Watch</label>' +
    '<select id="trk_metric" class="ed" style="width:150px">' + opts + "</select></div>" +
    '<div><label class="cc" style="display:block;font-size:11px">Target</label>' +
    '<input id="trk_target" class="ed" style="width:100px" placeholder="optional"></div>' +
    '<button class="primary" onclick="trkAdd()"><i class="ti ti-plus"></i> Track it</button>' +
    '<div class="cc" style="font-size:11px;max-width:320px">Without a target the value is ' +
    'recorded but nothing can be off track — there is nothing to be off.</div>' +
    "</div>";
}

async function trkLoad() {
  const box = document.getElementById("trk_body");
  if (box && !TRK.rows.length) box.innerHTML = '<div class="cc" style="padding:14px">Loading…</div>';
  try {
    const j = await (await fetch("/trackers" + _trkQs())).json();
    if (!j.ok) { if (box) box.innerHTML = '<div class="sresfail">' + esc(j.error || "failed") + "</div>"; return; }
    TRK.metrics = j.metrics || {};
    TRK.rows = j.rows || [];
  } catch (e) {
    if (box) box.innerHTML = '<div class="sresfail">' + esc(String(e)) + "</div>";
    return;
  }
  trkTabs();
  trkRender();
  trkBadge();
}

async function trkAdd() {
  const asin = (document.getElementById("trk_asin") || {}).value || "";
  const metric = (document.getElementById("trk_metric") || {}).value || "";
  const target = (document.getElementById("trk_target") || {}).value || "";
  if (!asin.trim()) { toast("Enter an ASIN."); return; }
  const j = await (await fetch("/trackers/watch" + _trkQs(), {
    method: "POST", headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ asin: asin, metric: metric, on: true, target: target })
  })).json();
  if (!j.ok) { toast(j.error || "Could not track that."); return; }
  const a = document.getElementById("trk_asin"); if (a) a.value = "";
  trkLoad();
}

async function trkSetTarget(asin, metric, value) {
  const j = await (await fetch("/trackers/watch" + _trkQs(), {
    method: "POST", headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ asin: asin, metric: metric, target: value })
  })).json();
  if (!j.ok) { toast(j.error || "Could not save that target."); return; }
  trkLoad();
}

async function trkStop(asin, metric) {
  await fetch("/trackers/watch" + _trkQs(), {
    method: "POST", headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ asin: asin, metric: metric, on: false })
  });
  trkLoad();
}

async function trkRefresh() {
  if (TRK.loading) return;
  TRK.loading = true;
  const btn = document.getElementById("trk_refresh");
  const was = btn ? btn.innerHTML : "";
  if (btn) { btn.innerHTML = '<i class="ti ti-loader"></i> Checking…'; btn.disabled = true; }
  try {
    const j = await (await fetch("/trackers/refresh" + _trkQs(), {
      method: "POST", headers: { "Content-Type": "application/json" },
      body: JSON.stringify({})
    })).json();
    if (!j.ok) toast(j.error || "Could not check.");
    else {
      // read AND stored, because they differ exactly when something failed and
      // one combined number would hide it.
      let msg = j.stored + " reading" + (j.stored === 1 ? "" : "s") +
                " from " + j.asins + " ASIN" + (j.asins === 1 ? "" : "s");
      if (j.read !== j.stored) msg += " (" + j.read + " read)";
      if (j.errors && j.errors.length) msg += " — " + j.errors.length + " problem(s)";
      toast(msg);
    }
  } catch (e) {
    toast(String(e));
  }
  if (btn) { btn.innerHTML = was; btn.disabled = false; }
  TRK.loading = false;
  trkLoad();
}

// ---- alerts ----------------------------------------------------------------

async function alertsLoad() {
  const box = document.getElementById("alr_body");
  if (box) box.innerHTML = '<div class="cc" style="padding:14px">Loading…</div>';
  let j;
  try {
    j = await (await fetch("/trackers/alerts" + _trkQs())).json();
  } catch (e) {
    if (box) box.innerHTML = '<div class="sresfail">' + esc(String(e)) + "</div>";
    return;
  }
  if (!j.ok) { if (box) box.innerHTML = '<div class="sresfail">' + esc(j.error || "failed") + "</div>"; return; }
  const cur = trkCur();
  if (!j.rows.length) {
    box.innerHTML = '<div class="card" style="padding:18px">' +
      '<div style="font-weight:600;margin-bottom:4px">Nothing is off target</div>' +
      '<div class="cc" style="font-size:12.5px">Only numbers with a target set can be off one. ' +
      'Anything that could not be read is shown as unknown on the Trackers screen rather than ' +
      "counted as fine here.</div></div>";
    trkBadge(0);
    return;
  }
  let html = '<div class="card" style="overflow-x:auto"><table class="stk-table"><thead><tr>' +
    "<th>Product</th><th>Tracker</th><th>Now</th><th>Target</th><th>Drift</th><th>Last read</th>" +
    "</tr></thead><tbody>";
  j.rows.forEach(function (r) {
    html += "<tr><td><div style=\"font-weight:600\">" + esc(r.asin) + "</div>" +
      '<div class="cc" style="font-size:11px">' + esc((r.name || "").slice(0, 60)) + "</div></td>" +
      "<td>" + esc(r.tracker) + "</td>" +
      "<td>" + trkFmt(r.value, r.kind, cur) + "</td>" +
      "<td>" + trkFmt(r.target, r.kind, cur) + "</td>" +
      "<td>" + trkDrift(r.drift) + "</td>" +
      '<td class="cc" style="font-size:11px">' + esc(r.last_at || "never") + "</td></tr>";
  });
  html += "</tbody></table></div>";
  box.innerHTML = html;
  trkBadge(j.count);
}

// The sidebar badge. Counted once, over all four trackers, because four badges
// would be four things to check and four chances to miss the one that mattered.
async function trkBadge(known) {
  const el = document.getElementById("alr_badge");
  if (!el) return;
  let n = known;
  if (n === undefined || n === null) {
    try {
      const j = await (await fetch("/trackers/alerts" + _trkQs())).json();
      n = j && j.ok ? j.count : 0;
    } catch (e) { n = 0; }
  }
  if (n > 0) { el.textContent = String(n); el.style.display = "inline-block"; }
  else { el.style.display = "none"; el.textContent = ""; }
  // A collapsed nav group has to repeat its children's badges, or shutting the
  // group would hide the alert.
  if (typeof navGroupBadges === "function") navGroupBadges();
}
